app/src/main/java/org/kiwix/kiwixmobile/webserver/ZimHostActivity.java

```diff
@@ -0,0 +1,517 @@
+      final Handler handler = new Handler();
+      handler.postDelayed(new Runnable() {
+        @Override
+        public void run() {
+          progressDialog.dismiss();
+          startService(ACTION_START_SERVER);
+        }
+      }, 7000);
+    });
+
+    builder.setTitle(getString(R.string.hotspot_dialog_title));
+    builder.setMessage(
+        getString(R.string.hotspot_dialog_message)
+    );
+    AlertDialog dialog = builder.create();
+    dialog.show();
+  }
+
+  void startService(String ACTION) {
+    serviceIntent.setAction(ACTION);
+    this.startService(serviceIntent);
+  }
+
+  void mobileDataDialog() {
+    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.O) {
+      AlertDialog.Builder builder = new AlertDialog.Builder(this, dialogStyle());
+
+      builder.setPositiveButton(this.getString(R.string.yes),
+          (dialog, id) -> openDataUsageActivity());
+      builder.setNegativeButton((android.R.string.no), (dialog, id) -> {
+        startHotspotDialog();
+      });
+      builder.setTitle(this.getString(R.string.mobile_data_enabled));
+      builder.setMessage(
+          this.getString(R.string.mobile_data_message) + "\n" + this.getString(
+              R.string.mobile_data_message_confirmation)
+      );
+      builder.setCancelable(false);
+      AlertDialog dialog = builder.create();
+      dialog.show();
+    }
+  }
+
+  @Override public void serverStarted(String ip) {
+    this.ip = ip;
+    serverTextView.setText(getString(R.string.server_started_message) + " " + this.ip);
+    startServerButton.setText(getString(R.string.stop_server_label));
+    startServerButton.setBackgroundColor(getResources().getColor(R.color.stopServer));
+    isServerStarted = true;
+  }
+
+  @Override public void serverStopped() {
+    serverTextView.setText(getString(R.string.server_textview_default_message));
+    startServerButton.setText(getString(R.string.start_server_label));
+    startServerButton.setBackgroundColor(getResources().getColor(R.color.greenTick));
+    isServerStarted = false;
+  }
+
+  @Override public void serverFailed() {
+    String message = String.format(getString(R.string.server_failed_message));
+    Toast.makeText(this, message, Toast.LENGTH_LONG).show();
+  }
+
+  @Override public void hotspotTurnedOn(WifiConfiguration wifiConfiguration) {
+
+    //Show an alert dialog for hotspot details
+    AlertDialog.Builder builder = new AlertDialog.Builder(this, dialogStyle());
+
+    ProgressDialog progressDialog =
+        ProgressDialog.show(this, getString(R.string.progress_dialog_starting_server), "",
+            true);
+    progressDialog.show();
+
+    builder.setPositiveButton(android.R.string.ok, (dialog, id) -> {
+      final Handler handler = new Handler();
+      handler.postDelayed(new Runnable() {
+        @Override
+        public void run() {
+          progressDialog.dismiss();
+          startService(ACTION_START_SERVER);
+          //webServerHelper.startServerHelper();
+        }
+      }, 2000);
```

Comment:
Why this delay?

Reply:
We have the callback for devices>Oreo when hotspot is started. But still in some cases it takes a second for the hotspot to get to the enabled state after it has been started.

Reply:
Do you get null if you get the IP inside the callback? If yes, call the method repeatedly until you get the IP after an interval of 1 second and if you don't get the IP in 20-30 seconds show a dialog saying the hotspot is not turned on.

Reply:
test

Reply:
Fixed. df9a28c3bb828f2e6e2e4780d37ab76f8afe6ddd